[ZEPPELIN-2647] Bypass auth logic when a user logins as admin role

yu74n opened the pull request:
### What is this PR for?
For administrator, make new admin role that assigned user can see all notebooks.

### What type of PR is it?
Improvement

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2647

### How should this be tested?
1. Set role name to use as admin through ZEPPELIN_OWNER_ROLE = <role name> or zeppelin.owner.role = <role name>.
Default role name is admin
2. Login as user who is not assigned as admin and create notebook.
3. Logout the user and login another user who is assigned as admin, open the created notebook.

### Questions:
* Does the licenses files need update? N
* Is there breaking changes for older versions? Y/N
* Does this needs documentation? Y

yu74n commented:
continued from #2525 #2416

1ambda commented:
Thanks for the contribution @yu74n 
Looks Good to Me. Let me test more cases and then left feedback soon.

felixcheung commented on conf/zeppelin-site.xml.template:
```diff
@@ -384,6 +384,12 @@
 </property>
 
 <property>
+  <name>zeppelin.owner.role</name>
```
I feel the property name might not be very intuitive, how about calling it `zeppelin.notebook.default.owner.username` or something similar?

yu74n replied:
Thank you for reviewing. It would be better, I think too. I'll fix it.

1ambda commented:
LGTM.

1ambda commented:
I wanted to merge now. But cound't find the build history. 

- https://builds.apache.org/job/zeppelin-pull-request/2960/ 

@yu74n if you don't mind close and reopen this issue to trigger jenkins?

yu74n commented:
@1ambda Sure

1ambda commented:
The 3/4 jobs failed. Please rebase this PR based on master. And Could you resolve them?

yu74n commented:
I rebased and ran test again, but this test is still failed.
https://travis-ci.org/yu74n/zeppelin/jobs/286508544

I think probably my PR doesn't affect the test.
I ran test several times, but failed tests are changed.

1ambda commented:
Hi, Thanks for rebasing and testing. But this recent PR passed the CI.
If you don't mind please re-run few times until successful and check the failure reasons? 

If we can verify they are a flaky test, we can merge, but without evidence, we can't do that.

![image](https://user-images.githubusercontent.com/4968473/31480264-c90d65e2-af57-11e7-9b21-47c6ecb0e42d.png)

yu74n commented:
test

yu74n commented:
Rerun all tests. 
'testPerUserScopedAction' is passed this time, but other 4 tests is failed.
These 4 tests was passed before.

yu74n commented:
All tests are passed, but NotebookTest has an error.
java.lang.RuntimeException: org.apache.thrift.transport.TTransportException: java.net.SocketException: Connection reset

yu74n commented:
and rerun the failed travis job manually, the job is passed.
https://travis-ci.org/yu74n/zeppelin/jobs/295676369

yu74n commented:
What do you think about that? @1ambda @felixcheung

yu74n commented:
@felixcheung if unset zeppelin.notebook.default.owner.username, set 'admin' as default value. 
https://github.com/apache/zeppelin/pull/2585/files#diff-179ea3a816210bb9904baabec1a552e4R716

yu74n commented:
I've changed default value is blank.
if set blank as zeppelin.notebook.default.owner.username, disable admin role.

felixcheung commented:
ok, could you explain how
> if set blank as zeppelin.notebook.default.owner.username, disable admin role.
?
I think also it'll be good to match template to the default value

yu74n commented:
if set blank as default owner username, apply blank('') to the admin role. but, I think, a user cannot set blank as login user name. so nearly equal as disable admin role.

felixcheung commented:
hmm, since this is around security perhaps we need to be more careful?
how about we add a check for the owner username is valid (at least not "") and not add to the admin role if it is not valid?

yu74n commented:
Actually, I thought so too when I explained how if the owner name is blank. I've fixed that.
I've changed config template to be disable admin role.
Can you review that again?
@felixcheung

felixcheung commented on conf/zeppelin-site.xml.template:
```diff
@@ -394,6 +394,12 @@
 <property>
+  <name>zeppelin.notebook.default.owner.username</name>
+  <value></value>
+  <description>Set owner role by default in private mode</description>
```
btw, did we say this isn't just for private mode any more?

yu74n replied:
This is for both public and private, then I've fixed it.

felixcheung commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/notebook/NotebookAuthorization.java:
```diff
@@ -293,24 +293,36 @@ public void setWriters(String noteId, Set<String> entities) {
+
+  private boolean isAdmin(Set<String> entities) {
+    String adminRole = conf.getString(ConfVars.ZEPPELIN_OWNER_ROLE);
+    if (adminRole.isEmpty()) {
```
please use org.apache.commons.lang.StringUtils IsBlank()

yu74n replied:
Thanks, fixed it.

felixcheung commented:
could you check if it's related? https://travis-ci.org/yu74n/zeppelin/builds/299517426

yu74n commented:
CI failed at testRunOnSelectionChange()
```bash
java.lang.AssertionError: Even if 'RunOnSelectionChange' is set as false, still can run the paragraph by pressing ENTER 
Expected: "My selection is 1"
but: was "My selection is 2"
```
This test is related to 'Run on selection change' checkbox in paragraph menu.
I think the reason why that test is failed would be timing error,  cause of the test code ran the paragraph before the checkbox was unchecked. Then I think the test is not related to my PR.

I rerun the CI job manually, all tests are passed.
https://travis-ci.org/yu74n/zeppelin/builds/299517426

felixcheung commented:
ok thanks

felixcheung commented:
merging if no more comment
